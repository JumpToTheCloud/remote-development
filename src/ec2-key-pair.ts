import * as fs from 'fs';
import * as path from 'path';
import { Annotations, CfnOutput, Stack } from 'aws-cdk-lib';
import { KeyPair } from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';

export interface KeypairProps {
  /**
   * This name will be used to the keypair name.
   * It will be a suffix added to "keypair-remote-development"
   * If not provided, we use the stack name as suffix
   */
  readonly keypairName?: string;

  /**
   * The name of the file of the public key we are going to upload to AWS.
   * This key will be used when we create the EC2 Instance
   *
   * The key must be stored in your computer in .ssh folder as usual.
   * You only need to provide the name of the file.
   *
   * @example
   *    Your file is ./ssh/mykey.pub
   *    You provide only the name of the file: "mykey.pub"
   */
  readonly sshPublicKeyName: string;
}

/**
 * @export
 * @class
 */
export class RemoteEnvironmentKeyPair extends Construct {
  /**
   * The key pair created
   */
  readonly key: KeyPair;

  /**
   * @param {Construct} scope
   * @param {string} id
   */
  constructor(scope: Construct, id: string, props: KeypairProps) {
    super(scope, id);

    const homeDirectory = process.env.HOME as string;
    const sshKeyPath = path.join(
      homeDirectory,
      `/.ssh/${props?.sshPublicKeyName}`
    );

    if (!fs.existsSync(sshKeyPath)) {
      Annotations.of(this).addError(
        `There is not ssh key file provided: ${homeDirectory}/.ssh/${props.sshPublicKeyName}. Please, create a new ssh key and save it in .shh directory`
      );
    }

    this.key = new KeyPair(this, 'ImportedKeyPair', {
      keyPairName: props.keypairName
        ? `keypair-remote-development-${props.keypairName}`
        : `keypair-remote-development-${Stack.of(this).stackName}`,
      publicKeyMaterial: fs.readFileSync(sshKeyPath).toString(),
    });

    new CfnOutput(this, 'OutputImportedKeyPair', {
      value: this.key.keyPairName,
      description: 'Name of the personal public key pair',
    });
  }
}
