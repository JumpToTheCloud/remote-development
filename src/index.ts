export * from './remote-ec2-instance';
