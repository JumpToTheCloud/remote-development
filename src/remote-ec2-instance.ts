import * as fs from 'fs';
import { Annotations, CfnOutput, Duration, Tags } from 'aws-cdk-lib';
import {
  ComparisonOperator,
  Metric,
  Stats,
  Unit,
} from 'aws-cdk-lib/aws-cloudwatch';
import {
  Ec2Action,
  Ec2InstanceAction,
} from 'aws-cdk-lib/aws-cloudwatch-actions';
import {
  BlockDeviceVolume,
  CfnEIP,
  Instance,
  InstanceType,
  MachineImage,
  Peer,
  Port,
  SecurityGroup,
  SubnetType,
  Vpc,
} from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';

import { RemoteEnvironmentKeyPair } from './ec2-key-pair';

/**
 * Ubuntu available product
 */
export enum UbuntuProduct {
  /**
   * Ubuntu Server
   */
  SERVER = 'server',

  /**
   * Minimal ubuntu server
   */
  'SERVER_MINIMAL' = 'server-minimal',
}

/**
 * Ubuntu version release
 */
export enum UbuntuRelease {
  /**
   * Release 22.04
   */
  'UBUNTU_22_04' = '22.04',

  /**
   * Release 20.04
   */
  'UBUNTU_20_04' = '20.04',

  /**
   * Release Focal
   */
  FOCAL = 'focal',

  /**
   * Release Bionic
   */
  BIONIC = 'bionic',
}

/**
 * Server Architecture
 */
export enum UbuntuArch {
  /**
   * X86 architecture
   */
  AMD64 = 'amd64',

  /**
   * ARM Architecture
   */
  ARM64 = 'arm64',
}

/**
 * Virtulization type
 */
export enum UbuntuVirtType {
  /**
   * Paravirtual type
   */
  PV = 'pv',

  /**
   * Hardware Virtual Machine type
   */
  HVM = 'hvm',
}

/**
 * EBS Volumen type attached to the instance
 */
export enum UbuntuVolumeType {
  /**
   * General Purpose SSD
   */
  EBS_GP2 = 'ebs-gp2',

  /**
   * Provisioned IOPS SSD Volumes
   */
  EBS_IO1 = 'ebs-io1',
}

/**
 * Ubuntu image configuration.
 * We use this to create the parameter store where the ubuntu ami id is located
 * The parameter is a conbination of this configuration
 */
export interface UbuntuImage {
  /**
   * Ubuntu product. For example: "server".
   */
  readonly product: UbuntuProduct;
  /**
   * Ubuntu release: For example: "20.04", "22.04", "focal", "bionic"
   */
  readonly release: UbuntuRelease;
  /**
   * Ubuntu architecture: For example: "amd64", "arm64"
   */
  readonly arch: UbuntuArch;
  /**
   * Ubuntu Virtualization Type: For example: "pv", "hvm"
   */
  readonly virt: UbuntuVirtType;
  /**
   * Ubuntu Volume Type: For example: "ebs-gp2", "ebs-io1"
   */
  readonly volume: UbuntuVolumeType;
}

export interface RemoteEnvironmentProps {
  /**
   * The VPC where the EC2 Instances is going to be launched.
   */
  readonly vpcId: string;
  /**
   * Name of the developer. This will be used to name the EC2 Instance.
   * Must be unique
   */
  readonly developer: string;
  /**
   * The name of the file of the public key we are going to upload to AWS.
   * This key will be used when we create the EC2 Instance
   *
   * The key must be stored in your computer in .ssh folder as usual.
   * You only need to provide the name of the file.
   *
   * @example
   *    Your file is ./ssh/mykey.pub
   *    You provide only the name of the file: "mykey.pub"
   */
  readonly sshPublicKeyName: string;
  /**
   * The number of Gigabites of the EBS Volumen attached to the instance
   */
  readonly ebsStorage?: number;

  /**
   * If you want to specify a version of ubuntu image to get the ami-id
   * you can use this configuration
   */
  readonly ubuntuImage?: UbuntuImage;

  /**
   * We use this when create an alarm to stop instance after this period.
   * By default is 15 minuts
   */
  readonly instanceStopMetricPeriod?: Duration;

  /**
   * EC2 Instance Type. For example: "t3.micro".
   * By default is "t3a.xlarge"
   */
  readonly instanceType?: string;
}

/**
 * Define a new Remote Environment in a EC2 instance for the given developer
 * @example
 * import { RemoteEnvironment} from '@jttc/remote-environment'
 *
 *  new RemoteEnvironment(stack, 'RemoteEnvironment', {
      developer: 'test',
      ebsStorage: 250,
      vpcId: 'vpc-xxxxxxxxxxx',
      sshPublicKeyName: 'id_ed25519.pub',
    });
 */
export class RemoteEnvironment extends Construct {
  private instance: Instance;
  /**
   * @param {Construct} scope
   * @param {string} id
   */
  constructor(scope: Construct, id: string, props: RemoteEnvironmentProps) {
    super(scope, id);

    const vpc = Vpc.fromLookup(this, 'VPC', {
      vpcId: props.vpcId,
    });

    const securityGroup = new SecurityGroup(this, 'SecurityGroup', {
      vpc: vpc,
      description: 'Allow ssh access to ec2 instance',
      allowAllOutbound: true,
      securityGroupName: `remote-environment-ec2-${props.developer}`,
    });

    securityGroup.addIngressRule(Peer.anyIpv4(), Port.tcp(22)),
      'Allow ssh connection';

    const keyPair = new RemoteEnvironmentKeyPair(this, 'KeyPair', {
      keypairName: props.developer,
      sshPublicKeyName: props.sshPublicKeyName,
    });

    let ubuntuImageParameter: string;

    if (props.ubuntuImage) {
      ubuntuImageParameter = `/aws/service/canonical/ubuntu/server/${props.ubuntuImage.release}/stable/current/${props.ubuntuImage.arch}/${props.ubuntuImage.virt}/${props.ubuntuImage.volume}/ami-id`;
    } else {
      ubuntuImageParameter = `/aws/service/canonical/ubuntu/server/${UbuntuRelease.UBUNTU_22_04}/stable/current/${UbuntuArch.AMD64}/${UbuntuVirtType.HVM}/${UbuntuVolumeType.EBS_GP2}/ami-id`;
    }

    const machineImage = MachineImage.fromSsmParameter(ubuntuImageParameter);
    const instanceType = props.instanceType ? props.instanceType : 't3a.xlarge';
    const ebsStorage = props.ebsStorage ? props.ebsStorage : 250;

    this.instance = new Instance(this, 'RemoteDevelopmentInstance', {
      vpc: vpc,
      instanceType: new InstanceType(instanceType),
      machineImage,
      keyPair: keyPair.key,
      securityGroup: securityGroup,
      vpcSubnets: {
        subnetType: SubnetType.PUBLIC,
      },
      blockDevices: [
        {
          deviceName: '/dev/sda1',
          volume: BlockDeviceVolume.ebs(ebsStorage),
        },
      ],
      associatePublicIpAddress: true,
    });

    Tags.of(this.instance).add('name', `remote-environment-${props.developer}`);

    const elasticIp = new CfnEIP(this, 'ElasticIp', {
      instanceId: this.instance.instanceId,
    });

    const metric = new Metric({
      namespace: 'AWS/EC2',
      metricName: 'CPUUtilization',
      statistic: Stats.MAXIMUM,
      period: props.instanceStopMetricPeriod
        ? props.instanceStopMetricPeriod
        : Duration.minutes(15),
      dimensionsMap: {
        InstanceId: this.instance.instanceId,
      },
      unit: Unit.PERCENT,
    });

    const alarm = metric.createAlarm(this, 'Ec2Alarm', {
      comparisonOperator: ComparisonOperator.LESS_THAN_OR_EQUAL_TO_THRESHOLD,
      threshold: 2,
      evaluationPeriods: 1,
      alarmName: `terminate-remote-environment-instance-15-minutes-${props.developer}`,
    });

    alarm.addAlarmAction(new Ec2Action(Ec2InstanceAction.STOP));

    new CfnOutput(this, 'PublicIp', {
      value: elasticIp.attrPublicIp,
      description: 'Public Ip',
    });
  }

  public addUserDate(scriptPath: string) {
    if (fs.existsSync(scriptPath)) {
      Annotations.of(this).addError(
        `There is not user data script in the path provider. Please, make sure there is file: ${scriptPath}`
      );
    }

    this.instance.addUserData(fs.readFileSync(scriptPath, 'utf-8').toString());
  }
}
