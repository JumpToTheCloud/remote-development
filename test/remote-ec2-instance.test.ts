import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { RemoteEnvironment } from '../src/remote-ec2-instance';

describe('Snapshot test validations', () => {
  const app = new App();
  const stack = new Stack(app, 'Stack', {
    env: {
      region: 'eu-west-1',
      account: '123456789012',
    },
  });

  new RemoteEnvironment(stack, 'RemoteEnvironment', {
    developer: 'test',
    ebsStorage: 250,
    vpcId: 'vpc-xxxxxxxxxxx',
    sshPublicKeyName: 'id_ed25519.pub',
  });

  it('should match the snapshot', () => {
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
