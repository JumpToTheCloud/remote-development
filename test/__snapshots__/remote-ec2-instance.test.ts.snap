// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test validations should match the snapshot 1`] = `
{
  "Outputs": {
    "RemoteEnvironmentKeyPairOutputImportedKeyPair566A0713": {
      "Description": "Name of the personal public key pair",
      "Value": {
        "Ref": "RemoteEnvironmentKeyPairImportedKeyPair0CBCD44B",
      },
    },
    "RemoteEnvironmentPublicIp93E7AF70": {
      "Description": "Public Ip",
      "Value": {
        "Fn::GetAtt": [
          "RemoteEnvironmentElasticIpBED6EA48",
          "PublicIp",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "SsmParameterValueawsservicecanonicalubuntuserver2204stablecurrentamd64hvmebsgp2amiidC96584B6F00A464EAD1953AFF4B05118Parameter": {
      "Default": "/aws/service/canonical/ubuntu/server/22.04/stable/current/amd64/hvm/ebs-gp2/ami-id",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::Image::Id>",
    },
  },
  "Resources": {
    "RemoteEnvironmentEc2Alarm903B6B96": {
      "Properties": {
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":automate:eu-west-1:ec2:stop",
              ],
            ],
          },
        ],
        "AlarmName": "terminate-remote-environment-instance-15-minutes-test",
        "ComparisonOperator": "LessThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "InstanceId",
            "Value": {
              "Ref": "RemoteEnvironmentRemoteDevelopmentInstance781276AB",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "CPUUtilization",
        "Namespace": "AWS/EC2",
        "Period": 900,
        "Statistic": "Maximum",
        "Threshold": 2,
        "Unit": "Percent",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "RemoteEnvironmentElasticIpBED6EA48": {
      "Properties": {
        "InstanceId": {
          "Ref": "RemoteEnvironmentRemoteDevelopmentInstance781276AB",
        },
      },
      "Type": "AWS::EC2::EIP",
    },
    "RemoteEnvironmentKeyPairImportedKeyPair0CBCD44B": {
      "Properties": {
        "KeyFormat": "pem",
        "KeyName": "keypair-remote-development-test",
        "KeyType": "rsa",
        "PublicKeyMaterial": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQD3F6tyPEFEzV0LX3X8BsXdMsQz1x2cEikKD",
      },
      "Type": "AWS::EC2::KeyPair",
    },
    "RemoteEnvironmentRemoteDevelopmentInstance781276AB": {
      "DependsOn": [
        "RemoteEnvironmentRemoteDevelopmentInstanceInstanceRole2268F6D1",
      ],
      "Properties": {
        "AvailabilityZone": "dummy1a",
        "BlockDeviceMappings": [
          {
            "DeviceName": "/dev/sda1",
            "Ebs": {
              "VolumeSize": 250,
            },
          },
        ],
        "IamInstanceProfile": {
          "Ref": "RemoteEnvironmentRemoteDevelopmentInstanceInstanceProfile58520179",
        },
        "ImageId": {
          "Ref": "SsmParameterValueawsservicecanonicalubuntuserver2204stablecurrentamd64hvmebsgp2amiidC96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "InstanceType": "t3a.xlarge",
        "KeyName": {
          "Ref": "RemoteEnvironmentKeyPairImportedKeyPair0CBCD44B",
        },
        "NetworkInterfaces": [
          {
            "AssociatePublicIpAddress": true,
            "DeviceIndex": "0",
            "GroupSet": [
              {
                "Fn::GetAtt": [
                  "RemoteEnvironmentSecurityGroup56BE87C1",
                  "GroupId",
                ],
              },
            ],
            "SubnetId": "s-12345",
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "remote-environment-test",
          },
        ],
        "UserData": {
          "Fn::Base64": "#!/bin/bash",
        },
      },
      "Type": "AWS::EC2::Instance",
    },
    "RemoteEnvironmentRemoteDevelopmentInstanceInstanceProfile58520179": {
      "Properties": {
        "Roles": [
          {
            "Ref": "RemoteEnvironmentRemoteDevelopmentInstanceInstanceRole2268F6D1",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "RemoteEnvironmentRemoteDevelopmentInstanceInstanceRole2268F6D1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "remote-environment-test",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "RemoteEnvironmentSecurityGroup56BE87C1": {
      "Properties": {
        "GroupDescription": "Allow ssh access to ec2 instance",
        "GroupName": "remote-environment-ec2-test",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "from 0.0.0.0/0:22",
            "FromPort": 22,
            "IpProtocol": "tcp",
            "ToPort": 22,
          },
        ],
        "VpcId": "vpc-12345",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
